src/input_common/gcadapter/gc_adapter.cpp

```diff
@@ -0,0 +1,369 @@
+        PadButton::PAD_TRIGGER_L};
+
+    if (adapter_controllers_status[port] != ControllerTypes::None) {
+        const u8 b1 = adapter_payload[1 + (9 * port) + 1];
+        const u8 b2 = adapter_payload[1 + (9 * port) + 2];
+
+        for (std::size_t i = 0; i < b1_buttons.size(); ++i) {
+            if (b1 & (1 << i)) {
+                pad.button |= static_cast<u16>(b1_buttons[i]);
+            }
+        }
+
+        for (std::size_t j = 0; j < b2_buttons.size(); ++j) {
+            if (b2 & (1 << j)) {
+                pad.button |= static_cast<u16>(b2_buttons[j]);
+            }
+        }
+
+        if (get_origin) {
+            pad.button |= PAD_GET_ORIGIN;
+        }
+
+        pad.stick_x = adapter_payload[1 + (9 * port) + 3];
+        pad.stick_y = adapter_payload[1 + (9 * port) + 4];
+        pad.substick_x = adapter_payload[1 + (9 * port) + 5];
+        pad.substick_y = adapter_payload[1 + (9 * port) + 6];
+        pad.trigger_left = adapter_payload[1 + (9 * port) + 7];
+        pad.trigger_right = adapter_payload[1 + (9 * port) + 8];
+    }
+    return pad;
+}
+
+void Adapter::PadToState(const GCPadStatus& pad, GCState& state) {
+    for (const auto& button : PadButtonArray) {
+        const u16 button_value = static_cast<u16>(button);
+        state.buttons.insert_or_assign(button_value, pad.button & button_value);
+    }
+
+    state.axes.insert_or_assign(static_cast<u8>(PadAxes::StickX), pad.stick_x);
+    state.axes.insert_or_assign(static_cast<u8>(PadAxes::StickY), pad.stick_y);
+    state.axes.insert_or_assign(static_cast<u8>(PadAxes::SubstickX), pad.substick_x);
+    state.axes.insert_or_assign(static_cast<u8>(PadAxes::SubstickY), pad.substick_y);
+    state.axes.insert_or_assign(static_cast<u8>(PadAxes::TriggerLeft), pad.trigger_left);
+    state.axes.insert_or_assign(static_cast<u8>(PadAxes::TriggerRight), pad.trigger_right);
+}
+
+void Adapter::Read() {
+    LOG_DEBUG(Input, "GC Adapter Read() thread started");
+
+    int payload_size_in, payload_size_copy;
+    std::array<u8, 37> adapter_payload;
+    std::array<u8, 37> adapter_payload_copy;
+    std::array<GCPadStatus, 4> pads;
+
+    while (adapter_thread_running) {
+        libusb_interrupt_transfer(usb_adapter_handle, input_endpoint, adapter_payload.data(),
+                                  sizeof(adapter_payload), &payload_size_in, 16);
+        payload_size_copy = 0;
+        // this mutex might be redundant?
+        {
+            std::lock_guard<std::mutex> lk(s_mutex);
+            std::copy(std::begin(adapter_payload), std::end(adapter_payload),
+                      std::begin(adapter_payload_copy));
+            payload_size_copy = payload_size_in;
+        }
+
+        if (payload_size_copy != sizeof(adapter_payload_copy) ||
+            adapter_payload_copy[0] != LIBUSB_DT_HID) {
+            LOG_ERROR(Input, "error reading payload (size: %d, type: %02x)", payload_size_copy,
```

Comment:
```suggestion
            LOG_ERROR(Input, "error reading payload (size: {}, type: {:02x})", payload_size_copy,
```

We use libfmt, so fmt specifiers should be used instead, otherwise this will crash at runtime if the error case ever occurs.

Reply:
Just for the record, it literally prints the `%d` rather than format it the way I had it 😅